### ISSUE TITLE
Drupal 7 upgrade path

### ISSUE THREAD

**Issue by bobchristenson** [backdrop-contrib/conditional_fields#17]
When migrating a Drupal 7 site to Backdrop, this module thows a ton of errors like:
`Notice: Undefined offset: 1 in conditional_fields_load_dependencies() (line 1308 of /modules/contrib/conditional_fields/conditional_fields.module).`

Let me know what other info you may need to track this down....

**Comment by bobchristenson** [comment]
Just for quick reference, here's the relevant function:

`/**
 * Loads all dependencies from the database.
 *
 * The result can be filtered by providing an entity type and a bundle name.
 */
function conditional_fields_load_dependencies($entity_type = NULL, $bundle = NULL) {
  // Use the advanced backdrop_static() pattern.
  static $dependencies;
  if (!isset($dependencies)) {
    $dependencies = &backdrop_static(__FUNCTION__);
  }

  if (!$dependencies) {
    $dependencies = array();
  }

  if (!isset($dependencies[$entity_type][$bundle])) {
    if (!empty($entity_type) && !empty($bundle)) {
      $dependencies[$entity_type][$bundle] = array();
    }
    $default_options = conditional_fields_dependency_default_options();

    $select = db_select('conditional_fields', 'cf')
      ->fields('cf', array('id', 'options', 'dependee', 'dependent'))
      ->orderBy('cf.dependent');

    $result = $select->execute();

    foreach ($result as $dependency) {
      $dependent_parts = explode('.', $dependency->dependent);
      $entity_type_match = (!$entity_type || ($entity_type && $dependent_parts[0] == $entity_type));
      $bundle_match = (!$bundle || ($bundle && $dependent_parts[1] == $bundle));
      if ($entity_type_match && $bundle_match) {
        // Make sure there is no Entity type or Bundle specified, or that the
        // specified Entity type or Bundle match the one that is returned
        // in this instance.
        $result_entity_type = $entity_type ? $entity_type : $dependent_parts[0];
        $result_bundle = $bundle ? $bundle : $dependent_parts[1];
  
        $options = unserialize($dependency->options);
        $options += $default_options;
  
        $dependencies[$result_entity_type][$result_bundle]['dependents'][$dependency->dependent][$dependency->id] = array(
          'dependee' => $dependency->dependee,
          'options' => $options,
        );
  
        $dependencies[$result_entity_type][$result_bundle]['dependees'][$dependency->dependee][$dependency->id] = array(
          'dependent' => $dependency->dependent,
          'options' => $options,
        );
      }
    }
  }

  if ($entity_type && isset($dependencies[$entity_type])) {
    if ($bundle && isset($dependencies[$entity_type][$bundle])) {
      return $dependencies[$entity_type][$bundle];
    }

    return $dependencies[$entity_type];
  }

  return $dependencies;
}`

**Comment by laryn** [comment]
@bobchristenson 
With `devel` installed, could you do a `dpm($dependent_parts);` after line 1306?

**Comment by bobchristenson** [comment]
I sure will once I’m able and I’ll report back here.

**Comment by bobchristenson** [comment]
Unfortunately it may be a while until I can test this...got pulled to another project for a while and it's not easy for me to load up the site and debug currently.  Will report back at a later date when I get back on this...

**Comment by laryn** [comment]
This may be a case where the D7 upgrade path was not put in place when the module was initially ported. The port happened slowly over time and I think it has decent functionality now as of Backdrop 1.20.3 but a D7 upgrade path is still needed before a formal full release.

**Comment by bobchristenson** [comment]
Finally back to this.  When I dpm($dependent_parts) I get: $dependent_parts[0] => 0

When I dpm($dependency) within the foreach, I'm seeing correct data from the db_select query...but the 'explode' doesn't make sense to me.  It seems like all the $dependency->dependent data is simple numbers (either 0 or a whole number).  What's the explode doing there?

**Comment by laryn** [comment]
@bobchristenson This is all related to making this module work with Backdrop's CMI and the data you have coming in must not be converted. The field instance IDs from D7 need to be converted into a series of values that allow us to know which field config is being referenced. So where D7 could use an int to reference the field instance ID, this Backdrop version stores something like `node.page.field_section` and the explode is to break that apart.

It's been a while since I looked at this but it sounds like in `conditional_fields_update_1000` we need to not just expand the length of the database field to allow for the larger "id" values but to convert the existing old numeric instance IDs to the new format, if we can.

**Comment by laryn** [comment]
@bobchristenson Can you test out `beta2` next time you're trying out the upgrade path and let me know if you see any issues on the D7 upgrade path?

**Comment by laryn** [comment]
By the way, I also re-tested the fix on https://github.com/backdrop-contrib/conditional_fields/issues/14 as I worked on this one. I set up a conditional field in Drupal 7 that changed another plain text field to required when it was checked. In Drupal 7 it added the asterisk but did not enforce the requirement (it would save it with the newly required field empty without a problem or even a warning). Then I upgraded to Backdrop and the fields were still configured correctly and the requirement was now enforced when adding/editing content. 👍 👍
